src/dra/7E4BC.c

```diff
@@ -1235,7 +1235,435 @@ Primitive* func_80121F58(bool arg0, s32 arg1, Primitive* arg2, s16 facingLeft) {
 }
 
 // spawns mist (player transform)
-INCLUDE_ASM("dra/nonmatchings/7E4BC", EntityMist);
+void EntityMist(Entity* self) {
+    Primitive* prim;
+    Primitive* mistPrim;
+    mistStruct* mistStruct;
+    s16 xVar;
+    s16 yVar;
+    s16 angle;
+    s16 magnitude;
+    s16 angle2;
+    s16 angle3;
+    s16 xSize;
+    s16 ySize;
+    s16 baseSize;
+    s16 index;
+    s16 xVar2;
+    s16 yVar2;
+    s16 xVar3;
+    s16 yVar3;
+    s32 primColor1;
+    s32 primColor2;
+    s32 primColor3;
+    s32 primColor4;
+    s16 xVar4;
+    s16 yVar4;
+    s32 velX;
+    s32 velY;
+    s32 cosine;
+    s32 sine;
+    s32 i;
+    s32 j;
+    s16 xVar5;
+    s16 yVar5;
+
+    velX = PLAYER.velocityX >> 1;
+    velY = PLAYER.velocityY >> 1;
+    primColor1 = D_801381E4 + (rsin(D_801381D4) >> 8);
+    primColor2 = D_801381E8 + (rsin(D_801381D8) >> 8);
+    primColor3 = D_801381EC + (rsin(D_801381DC) >> 8);
+    primColor4 = D_801381F0 + (rsin(D_801381E0) >> 8);
+    D_801381D4 += 0x20;
+    D_801381D8 -= 0x20;
+    D_801381DC += 0x20;
+    D_801381E0 -= 0x20;
+    self->ext.mist.timer++;
+
+    self->facingLeft = PLAYER.facingLeft;
+    if (PLAYER.animSet == 0xF) {
+        self->facingLeft = 0;
+    }
+    switch (self->step) {
+    case 0:
+        // Weird that we set FLAG_DEAD here.
+        self->params = FLAG_UNK_4000 | FLAG_UNK_1000 | FLAG_UNK_800 |
+                       FLAG_UNK_400 | FLAG_DEAD;
+        EntityPlayerBlinkWhite(self);
+        if (self->primIndex == -1) {
+            goto block_147;
+        }
+        prim = &g_PrimBuf[self->primIndex];
+        mistPrim = D_801381F4;
+        for (i = 0; i < 8; i++, mistPrim++) {
+            *mistPrim = *prim;
+            prim = prim->next;
+        }
+        FreePrimitives(self->primIndex);
+        self->step = 0;
+        self->primIndex = AllocPrimitives(PRIM_GT4, 80);
+        if (self->primIndex == -1) {
+            goto block_147;
+        }
+// Not sure why PSP masks to 7FFF. Might be worth digging into.
+// There might be some special magic in how the systems differ.
+// I don't think this is a real ifdef.
+#ifdef VER_PSP
+        D_801381D4 = rand() & 0x7FFF;
+        D_801381D8 = rand() & 0x7FFF;
+        D_801381DC = rand() & 0x7FFF;
+        D_801381E0 = rand() & 0x7FFF;
+#else
+        D_801381D4 = rand();
+        D_801381D8 = rand();
+        D_801381DC = rand();
+        D_801381E0 = rand();
+#endif
```

Comment:
Nit: might be cleaner to extract a `rand_mask` const as in https://decomp.me/scratch/G3ECA

Reply:
I think it would be better not having a constant here? At most a `const int RandMax = 0x7FFF` but it is so self-contained and context-explicit that I would keep it as it is. IMHO.